### DISCUSSION TITLE
How can I put $ " «... key in defsrc to configure a typematrix bépo layout?

### DISCUSSION THREAD

**Discussion by stephane-klein** [kmonad/kmonad#611] [Q&A] [answered]
Hi,

I have a [TypeMatrix](http://typematrix.com/) [Bépo](http://bepo.fr/wiki/Accueil) keyboard.

I have this `defsrc`:

```
(defcfg
  input  (device-file "/dev/input/by-id/usb-TypeMatrix.com_USB_Keyboard-event-kbd")
  output (uinput-sink "kmonad_output_kb")
)

(defsrc
  esc   f1    f2    f3   f4   f5   del     f6    f7   f8    f9    f10   f11    f12   -
  $     "     «     »    (    )   bspc     @     +    -     /     *     =      %
  tab   b     é     p    o    è    -       ^     v    d     l     j     z      w
  lsft  a     u     i    e    ,    ret     c     t    s     r     n     m      rsft
  -     à     y     x    .    k    -       '     q    g     h     f     ç      -
  lctl  -     -     -              spc                      home  up    end    rctl  pgup
  fn    lmet  lalt                                    ralt  left  down  right  -     pgdown
)
```

When I check this configuration file, I have this error:

```
# /home/stephane/bin/kmonad-0.4.1-linux -d bepo-typematrix.kbd
kmonad-0.4.1-linux: Parse error at 8:3:
  |
8 |   $     "     «     »    (    )   bspc     @     +    -     /     *     =      %
  |   ^
unexpected '$'
expecting ')' or keycode
```

I do not find these characters in [`Keycode.hs`](https://github.com/kmonad/kmonad/blob/master/src/KMonad/Keyboard/Keycode.hs) then I don't know how to define my layout :thinking:.

Can you help me?

Best regards,
Stéphane

**Comment by MuhammedZakir** [comment, marked as answer]
Keycode/keyname is different than what's printed on the keyboard. To find the keycode,

1. use a config like

```clojure
(defcfg
  input  (device-file "/dev/input/by-id/usb-TypeMatrix.com_USB_Keyboard-event-kbd")
  output (uinput-sink "kmonad_output_kb")
  fallthrough false
)

(defsrc
  lctl  c  ;; for killing proccess via Ctrl+c
)

(deflayer main
  lctl  c
)
```

2. Run `kmonad -l debug /path/to/config` from terminal.

3. Press keys of the captured keyboard, and KMonad will log events it got, which will contain keycode (between `<` and `>`).

Edit: correction: forgot to add a `deflayer`.

---

Important: you can only put "keycode" in `defsrc`. For unicode characters, they are strictly OS-specific functionality. You can use them in `deflayer` if KMonad has implemented support for them in that OS. Currently, only Linux is supported.

**Reply by stephane-klein** [reply]
@MuhammedZakir I have this error:

```
[root@nuc-i7 chezmoi]# cat config.kbd
(defcfg
  input  (device-file "/dev/input/by-id/usb-TypeMatrix.com_USB_Keyboard-event-kbd")
  output (uinput-sink "kmonad_output_kb")
  fallthrough false
)

(defsrc
  lctl  c  ;; for killing proccess via Ctrl+c
)

[root@nuc-i7 chezmoi]# /home/stephane/bin/kmonad-0.4.1-linux -l debug config.kbd
kmonad-0.4.1-linux: Missing at least 1 block of type: deflayer
```

:thinking:

**Reply by MuhammedZakir** [reply]
Sorry, forgot to add a `deflayer`. I have corrected it.
